registry/storage/driver/s3-aws/s3.go

```diff
@@ -1380,61 +1443,83 @@ func (w *writer) Write(p []byte) (int, error) {
 
 	var n int
 
-	for len(p) > 0 {
-		// If no parts are ready to write, fill up the first part
-		if neededBytes := int(w.driver.ChunkSize) - len(w.readyPart); neededBytes > 0 {
-			if len(p) >= neededBytes {
-				w.readyPart = append(w.readyPart, p[:neededBytes]...)
-				n += neededBytes
-				p = p[neededBytes:]
-			} else {
-				w.readyPart = append(w.readyPart, p...)
-				n += len(p)
-				p = nil
-			}
+	reader := bytes.NewReader(p)
+
+	for {
+		// NOTE(milosgajdos): we do some seemingly unsafe conversions
+		// from int64 to int in this for loop. These are fine as the
+		// offset returned from buffer.ReadFrom can only ever be
+		// maxChunkSize large which fits in to int. The reason why
+		// we return int64 is to play nice with Go interfaces where
+		// the buffer implements io.ReaderFrom interface.
+
+		// fill up the ready parts buffer
+		offset, err := w.ready.ReadFrom(reader)
+		if err != nil {
+			return 0, err
 		}
+		n += int(offset)
 
-		if neededBytes := int(w.driver.ChunkSize) - len(w.pendingPart); neededBytes > 0 {
-			if len(p) >= neededBytes {
-				w.pendingPart = append(w.pendingPart, p[:neededBytes]...)
-				n += neededBytes
-				p = p[neededBytes:]
-				err := w.flushPart()
-				if err != nil {
-					w.size += int64(n)
-					return n, err
-				}
-			} else {
-				w.pendingPart = append(w.pendingPart, p...)
-				n += len(p)
-				p = nil
+		// we've read all bytes
+		if reader.Len() == 0 {
+			break
+		}
+
+		// try filling up the pending parts buffer
+		offset, err = w.pending.ReadFrom(reader)
+		if err != nil {
+			return 0, err
+		}
+		n += int(offset)
+
+		// we didn't fill up pending buffer, flush
+		if w.pending.Len() <= w.pending.Cap() {
```

Comment:
`len(a) <= cap(a)` is an invariant, therefore this condition is always true. It also does not match the conditions under which the old code would flush the buffer. The logic of the old code was "if there is enough remaining data in `p` to fill the pending buffer to capacity, fill the pending buffer and flush."
```suggestion
		// we filled up pending buffer, flush
		if w.pending.Len() == w.pending.Cap() {
```

Reply:
Yah. Great catch.